csrc/val_graph.cpp

```diff
@@ -0,0 +1,520 @@
+  NVF_ERROR(val_group);
+  return unique_definitions_.find(val_group) != unique_definitions_.end();
+}
+
+bool ValGraph::hasUses(const ValGroup& val_group) const {
+  NVF_ERROR(val_group);
+  return unique_uses_.find(val_group) != unique_uses_.end();
+}
+
+std::string ValGraph::toString() const {
+  std::stringstream ss;
+  ss << "IdGraph { \n";
+  ss << "Disjoint Ids:\n"
+     << idGroupsString(*this, 1) << "\n\nDisjoint Expression groups:\n"
+     << exprGroupsString(*this, 1) << std::endl;
+  ss << " } IdGraph\n" << std::endl;
+  return ss.str();
+}
+
+std::vector<std::vector<Val*>> ValGraph::isTrivialExpr(Expr* expr) {
+  std::vector<std::vector<Val*>> mapped_ids;
+  if (auto merge = dynamic_cast<Merge*>(expr)) {
+    if (merge->inner()->extent()->isOneInt()) {
+      mapped_ids.push_back({merge->outer(), merge->out()});
+    }
+    if (merge->outer()->extent()->isOneInt()) {
+      mapped_ids.push_back({merge->inner(), merge->out()});
+    }
+  } else if (auto split = dynamic_cast<Split*>(expr)) {
+    if (split->factor()->isOneInt() && split->startOffset()->isZeroInt() &&
+        split->stopOffset()->isZeroInt()) {
+      if (split->innerSplit()) {
+        mapped_ids.push_back({split->in(), split->outer()});
+      } else {
+        mapped_ids.push_back({split->in(), split->inner()});
+      }
+    }
+  } else if (auto swizzle = dynamic_cast<Swizzle2D*>(expr)) {
+    if (swizzle->swizzleType() == Swizzle2DType::NoSwizzle ||
+        swizzle->swizzleMode() == SwizzleMode::NoSwizzle) {
+      mapped_ids.push_back({swizzle->inX(), swizzle->outX()});
+      mapped_ids.push_back({swizzle->inY(), swizzle->outY()});
+    }
+  }
+  return mapped_ids;
+}
+
+bool ValGraph::exprAttributesMatch(Expr* first, Expr* second) {
+  if (first == nullptr || second == nullptr) {
+    return false;
+  }
+
+  if (typeid(*first) != typeid(*second)) {
+    return false;
+  }
+
+  if (first->isA<Split>()) {
+    auto first_split = first->as<Split>();
+    auto second_split = second->as<Split>();
+    if (!first_split->factor()->sameAs(second_split->factor()) ||
+        first_split->innerSplit() != second_split->innerSplit() ||
+        !first_split->startOffset()->sameAs(second_split->startOffset()) ||
+        !first_split->stopOffset()->sameAs(second_split->stopOffset())) {
+      return false;
+    }
+  }
+
+  if (first->isA<Swizzle2D>()) {
+    auto first_swizzle = first->as<Swizzle2D>();
+    auto second_swizzle = second->as<Swizzle2D>();
+    if (first_swizzle->swizzleMode() != second_swizzle->swizzleMode() ||
+        first_swizzle->swizzleType() != second_swizzle->swizzleType()) {
+      return false;
+    }
+  }
+
+  if (first->isA<Resize>()) {
+    if (!first->as<Resize>()->leftExpand()->sameAs(
+            second->as<Resize>()->leftExpand())) {
+      return false;
+    }
+
+    if (!first->as<Resize>()->rightExpand()->sameAs(
+            second->as<Resize>()->rightExpand())) {
+      return false;
+    }
+  }
+
+  return true;
+}
```

Comment:
Should this go to `IdModel`? I think `ValGraph` should provide an API for mapping `Val`s and `Expr`s when requested, but should not look into the `Expr`s and check if they match. For example, if I have
```
b2 = add(b0, b1)
b3 = or(b0, b1)
```
Then I should be able to map the `add` and `or` without checking their attributes match, because for bool `add` is equivalent to `or`, although they have different `BinaryOpType`.

In general, `ValGraph` is a data structure used to model equivalence relationship with graph structure, however, it does not define what the relationship is. The users of `ValGraph`, here `IdModel`, defines the equivalence relationship.

Reply:
Agreed. Let me rewrite it.

Reply:
Actually, `isTrivialExpr` is not used in this PR. I'll remove it for now and will address it in a coming PR where it's indeed used.